src/main.rs

```diff
@@ -0,0 +1,223 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::{io, sync::Arc};
+
+use agnostic::{shutdown_enclave_stream, EnclaveAddr, EnclaveStream, DEFAULT_DEST_ADDR};
+use anyhow::{Context, Result};
+use bytes::BytesMut;
+use clap::{command, Parser};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream},
+    sync::Semaphore,
+};
+use tracing::{debug, error, info, metadata::LevelFilter};
+use tracing_subscriber::EnvFilter;
+
+use crate::agnostic::{connect_to_enclave, parse_enclave_addr};
+
+#[cfg(not(feature = "mock-vsock"))]
+mod agnostic {
+    use std::net::Shutdown;
+
+    use anyhow::{anyhow, Result};
+    use tokio_vsock::{VsockAddr, VsockStream};
+
+    pub type EnclaveStream = VsockStream;
+    pub type EnclaveAddr = VsockAddr;
+
+    pub const DEFAULT_DEST_ADDR: &str = "4:8443";
+
+    pub fn parse_enclave_addr(address: &str) -> Result<EnclaveAddr> {
+        let mut address_split = address.split(":");
+        let cid = address_split
+            .next()
+            .ok_or(anyhow!("missing cid from vsock addr: {address}"))?
+            .parse()?;
+        let port = address_split
+            .next()
+            .ok_or(anyhow!("missing port from vsock addr: {address}"))?
+            .parse()?;
+        Ok(VsockAddr::new(cid, port))
+    }
+
+    pub async fn connect_to_enclave(address: EnclaveAddr) -> Result<EnclaveStream> {
+        Ok(VsockStream::connect(address.cid(), address.port()).await?)
+    }
+
+    pub async fn shutdown_enclave_stream(stream: &mut EnclaveStream) {
+        stream.shutdown(Shutdown::Both).ok();
+    }
+}
+
+#[cfg(feature = "mock-vsock")]
+mod agnostic {
+    use std::{net::SocketAddr, str::FromStr};
+
+    use anyhow::{Context, Result};
+    use tokio::{io::AsyncWriteExt, net::TcpStream};
+
+    pub type EnclaveStream = TcpStream;
+    pub type EnclaveAddr = SocketAddr;
+
+    pub const DEFAULT_DEST_ADDR: &str = "127.0.0.1:9443";
+
+    pub fn parse_enclave_addr(address: &str) -> Result<EnclaveAddr> {
+        Ok(SocketAddr::from_str(address).context("error parsing destination address")?)
+    }
+
+    pub async fn connect_to_enclave(address: EnclaveAddr) -> Result<EnclaveStream> {
+        Ok(TcpStream::connect(address).await?)
+    }
+
+    pub async fn shutdown_enclave_stream(stream: &mut EnclaveStream) {
+        stream.shutdown().await.ok();
+    }
+}
+
+/// Relays TCP connections from IPV4/IPV6 to VSOCK.
+#[derive(Parser, Debug)]
+#[command(version)]
+struct Cli {
+    /// Buffer size to use when reading/writing data between peers
+    #[arg(long, default_value_t = 8192)]
+    buffer_size: usize,
+
+    /// IPV4/IPV6 address/port to listen on
+    #[arg(short = 's', long, default_value = "0.0.0.0:8443")]
```

Comment:
does this need to be 0.0.0.0 or is 127.0.0.1 ok?

Reply:
In most scenarios, this relay would be open to the public. Are you suggesting `127.0.0.1` as a sensible default for testing purposes?

Reply:
i have a slight preference for the default to be a specific address to avoid accidentally listening on all interfaces if that's not the intention, but nbd either way